### PR TITLE
2FA with TOTP

### LINKED ISSUE #2201: [FR] Two-Factor Authentication
Adding another layer of security to the authentication system would be a great addition (OTP).

**Do you want to develop this?**
No, but we are open to create an open-source bounty.

### PR DESCRIPTION
**Pull request by matmair**
This PR adds MFA per TOTP.
- It changes the permission functions to also handle app names with underscores
- Adds TOTP and static backup tokens into the autentication chain
- Adds a global setting to require MFA
- Adds a user setting section to handle MFA settings
- Refactors security details in the user settings

Closes #2201

<a href="https://gitpod.io/#https://github.com/inventree/InvenTree/pull/2221"></a>

### CODE REVIEW

File: InvenTree/common/models.py [hunk 1]
```diff
@@ -853,6 +853,12 @@ class InvenTreeSetting(BaseInvenTreeSetting):
             'default': '',
             'choices': settings_group_options
         },
+        'LOGIN_ENFORCE_MFA': {
+            'name': _('Enforce MFA'),
+            'description': _('Users must use multifaktor security.'),
```
Comment (SchrodingersGat) [inline InvenTree/common/models.py:858]: "multifaktor" -> "multifactor"

File: InvenTree/users/models.py [hunk 2]
```diff
@@ -377,14 +393,22 @@ def get_permission_object(permission_string):
...
+        model = model[-1:][0]
 
         try:
             content_type = ContentType.objects.get(app_label=app, model=model)
             permission = Permission.objects.get(content_type=content_type, codename=perm)
         except ContentType.DoesNotExist:
             logger.warning(f"Error: Could not find permission matching '{permission_string}'")
             permission = None
+        except Exception as _e:
+            print(_e)
```
Comment (SchrodingersGat) [inline InvenTree/users/models.py:411]: probably should throw `_e` into a `logger.error` call

**Comment by SchrodingersGat** [comment]
This is really cool @matmair

Can you please also submit a docs PR for this one?

File: InvenTree/users/models.py (hunk 2 above)
Comment (matmair) [reply InvenTree/users/models.py:411]: this is removed in the next commit; only needed it for debug

File: InvenTree/common/models.py (hunk 1 above)
Comment (matmair) [reply InvenTree/common/models.py:858]: my german is showing 😇

**Comment by SchrodingersGat** [comment]
Some rendering issues on the user email settings page:

image

In addition to fixing the translation issue, it would be nice to put this inside a warning alert block

**Comment by SchrodingersGat** [comment]
Change spelling of "faktor" to "factor" in multiple locations

image

**Comment by matmair** [comment]
> 
> 
> This is really cool @matmair
> 
I hope so, it is my 3rd attempt. The available libs do not make me happy. I might fork something soon.
> Can you please also submit a docs PR for this one?
Already in the works, need to clean up the user section first

**Comment by SchrodingersGat** [comment]
image

**Comment by SchrodingersGat** [comment]
image

Once tokens are generated, there is no way to get back to the main page? Maybe a "return to settings" button?

**Comment by matmair** [comment]
@SchrodingersGat should be good to go once the test are through

**Comment by matmair** [comment]
@SchrodingersGat could you do a testrun before merge? I had some not reproducible problems on a already existing system

**Comment by matmair** [comment]
@SchrodingersGat ok fiexed it now

**Comment by SchrodingersGat** [comment]
I'm getting the following error when trying to run locally:

image

**Comment by SchrodingersGat** [comment]
This button is broken in the latest commit:

image

**Comment by SchrodingersGat** [comment]
I think it should be:

```html
<a href='/accounts/two_factor/setup' type='btn btn-primary' role='button'></a>
```

**Comment by SchrodingersGat** [comment]
Functionality wise, it seems to work well :)

**Comment by SchrodingersGat** [comment]
One small issue - in the google authenticator app, after scanning the QR code it adds an entry for "example.com"

Can we include a name for the entry in the QR code? It would very useful to enumerate the entry as the custom instance name, for example, or at the very least "InvenTree"

**Comment by matmair** [comment]
> One small issue - in the google authenticator app, after scanning the QR code it adds an entry for "example.com"
> 
> Can we include a name for the entry in the QR code? It would very useful to enumerate the entry as the custom instance name, for example, or at the very least "InvenTree"

It actually uses the name setting for the current domain as set in the site framework - I will add that to the docs

**Comment by SchrodingersGat** [comment]
> It actually uses the name setting for the current domain as set in the site framework - I will add that to the docs

Is this specified in the django settings file? Or somewhere else?

**Comment by SchrodingersGat** [comment]
Recent merge has broken a tag on the settings page:

image

**Comment by SchrodingersGat** [comment]
@matmair some other issues I think I have found:

### Admin Access

It appears that users can still access anything under `/admin/` without 2FA, even if 2FA is being enforced.

### Lockout

I managed to lock myself out of my dev server, with the following process:

1. Set up a 2FA account for the current user. Sign in using password and 2FA authenticator app
1. Turn on "enforce 2FA" option
1. Remove 2FA factors for the current user (in the settings window)
1. Log out
1. Presented with a new screen to setup 2FA.

At this point, setting up a new authenticator / 2FA session, any tokens / codes I put in fail. I had to manually remove the `ENFORCE_2FA` setting in the database to be able to log back in...

**Comment by matmair** [comment]
> ### Admin Access
> 
> It appears that users can still access anything under `/admin/` without 2FA, even if 2FA is being enforced.

This is actually so you cant lock yourself out.
> ### Lockout
> 
> I managed to lock myself out of my dev server, with the following process:
> ...
> At this point, setting up a new authenticator / 2FA session, any tokens / codes I put in fail. I had to manually remove the `ENFORCE_2FA` setting in the database to be able to log back in...
Will try to reproduce and fix

**Comment by SchrodingersGat** [comment]
> This is actually so you cant lock yourself out.

Is this not a security issue though?

I'm not going to reject the PR due to this, I'd like to understand your reasoning, though. And you do have a *lot* more knowledge of this sort of thing than I do.

**Comment by SchrodingersGat** [comment]
@matmair let me know when you're happy for review again :)

**Comment by SchrodingersGat** [comment]
@matmair ready to look at this again, or more work to be done?

**Comment by matmair** [comment]
> 
> 
> @matmair ready to look at this again, or more work to be done?

@SchrodingersGat I am still waiting for a response from a college that tests the security of the whole thing for me.

**Comment by SchrodingersGat** [comment]
@matmair any further movement on this one?

**Comment by matmair** [comment]
@SchrodingersGat have been working on other stuff for the last while, will look through my notes and make the needed changes.

**Comment by SchrodingersGat** [comment]
No pressure; you have been contributing a lot of really great stuff lately!

**Comment by matmair** [comment]
@SchrodingersGat could you check if you can reproduce the error with adding a factor?
I will add a management command to remove MFA from an account. Today or tomorrow, should be ready then.

**Comment by SchrodingersGat** [comment]
@matmair looking good now, cannot replicate that error.

The one suggestion I have now is that the "setup MFA" page needs a button to return to the settings page, otherwise it is a bit of a dead end:

image

**Comment by matmair** [comment]
@SchrodingersGat ready when the checks have passed

**Pull request merged**